File: python/src/adeu/pagination.py
# FILE: src/adeu/pagination.py
"""
Stateless paginator for projected DOCX Markdown.

Splits the post-projection Markdown body into virtual pages of <= PAGE_TARGET_CHARS,
respecting CriticMarkup boundaries, table integrity, and footnote sections.

Pagination is computed deterministically on every call — no caching, no state.
The Structural Appendix (if present) is appended to every page.

Used by the read_docx MCP tool's pagination mode and indirectly by outline mode
(which needs body_pages + body_page_offsets to map heading positions to page numbers).
"""

import re
from dataclasses import dataclass, field
from typing import Dict, List, Literal, Tuple, Union

from adeu.utils.docx import PAGE_BREAK_TOKEN

PAGE_TARGET_CHARS = 19_000
PAGE_RANGE_MAX_PAGES = 8
APPENDIX_MARKER = "<!-- READONLY_BOUNDARY_START -->"

_PAGE_RANGE_RE = re.compile(r"^\s*(\d+)\s*-\s*(\d+)\s*$")


def parse_page_arg(
    page: Union[int, str, None],
) -> Tuple[Literal["single", "range", "all"], Union[int, Tuple[int, int], None]]:
    """
    Parses and validates a page parameter (`page`).

    Args:
        page: Page argument supplied as int, str, or None.

    Returns:
        Tuple of (kind, value):
          - ("single", int_page_number)
          - ("range", (start_page, end_page))
          - ("all", None)

    Raises:
        ValueError: if page parameter is an invalid string, 0, or negative integer.
    """
    if page is None:
        return ("single", 1)

    if isinstance(page, int):
        if page < 1:
            raise ValueError(
                f"Invalid page parameter: '{page}'. Provide a positive integer, page range (e.g. '2-6'), or 'all'."
            )
        return ("single", page)

    if isinstance(page, str):
        s_page = page.strip()
        if not s_page:
            raise ValueError(
                f"Invalid page parameter: '{page}'. Provide a positive integer, page range (e.g. '2-6'), or 'all'."
            )

        if s_page.lower() == "all":
            return ("all", None)

        range_match = _PAGE_RANGE_RE.match(s_page)
        if range_match:
            start_p = int(range_match.group(1))
            end_p = int(range_match.group(2))
            if start_p < 1 or end_p < 1:
                raise ValueError(
                    f"Invalid page parameter: '{page}'. Provide a positive integer, page range (e.g. '2-6'), or 'all'."
                )
            return ("range", (start_p, end_p))

        try:
            val = int(s_page)
        except ValueError:
            raise ValueError(
                f"Invalid page parameter: '{page}'. Provide a positive integer, page range (e.g. '2-6'), or 'all'."
            ) from None

        if val < 1:
            raise ValueError(
                f"Invalid page parameter: '{page}'. Provide a positive integer, page range (e.g. '2-6'), or 'all'."
            )
        return ("single", val)

    raise ValueError(
        f"Invalid page parameter: '{page}'. Provide a positive integer, page range (e.g. '2-6'), or 'all'."
    )


# CriticMarkup open-token -> close-token. Order matters in scanning: longer/more-specific
# tokens are checked first. All four open tokens here are 3 chars, so dict order is fine.
_CRITIC_TOKENS: Dict[str, str] = {
    "{++": "++}",
    "{--": "--}",
    "{==": "==}",
    "{>>": "<<}",
}

# Note: this regex may match the literal string "Chg:42" appearing inside user-authored
# document text (e.g. quoted code, sample data). In practice this is vanishingly rare
# in real documents and accepting the false-positive risk is cheaper than parsing
# the meta-block syntax. See plan Concern 4.
_CHG_ID_PATTERN = re.compile(r"\bChg:(\d+)\b")

# Single-pass scanner event stream for _split_on_safe_paragraph_breaks: every
# CriticMarkup token plus paragraph-break newline runs. At any string position
# at most ONE alternative can match (opens start '{', each close starts with a
# distinct char, breaks start '\n'), and re.finditer consumes non-overlapping
# matches left to right — together this reproduces the historical
# char-by-char scan exactly (see tests/test_pagination_scanner_equivalence.py).
_SCAN_TOKEN_RE = re.compile(r"\{\+\+|\{--|\{==|\{>>|\+\+\}|--\}|==\}|<<\}|\n{2,}")


@dataclass(slots=True)
class PageInfo:
    page: int
    total_pages: int
    has_next: bool
    has_prev: bool
    tracked_change_count: int
    page_content: str


@dataclass(slots=True)
class PaginationResult:
    pages: List[PageInfo]
    total_pages: int
    # Pre-appendix-injection page bodies, in order. Used by outline.py to compute
    # heading -> page mapping without contending with appendix offsets.
    body_pages: List[str] = field(default_factory=list)
    # Start offset (in the original markdown_body string) at which each body page
    # begins. body_page_offsets[i] corresponds to body_pages[i]. Always 1 entry per
    # body_page; body_page_offsets[0] is always 0.
    body_page_offsets: List[int] = field(default_factory=list)


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------


def split_structural_appendix(markdown: str) -> Tuple[str, str]:
    """
    Splits the projected Markdown into (body, appendix).

    The appendix begins at the line containing APPENDIX_MARKER and includes
    everything from that line onward. The body is everything before it, with
    trailing whitespace stripped so we don't carry a "\\n\\n---\\n\\n" tail
    that domain.py emits before the marker.
    """
    if not markdown:
        return "", ""

    idx = markdown.find(APPENDIX_MARKER)
    if idx == -1:
        return markdown, ""

    # Walk back to the start of the line containing the marker.
    line_start = markdown.rfind("\n", 0, idx) + 1  # rfind returns -1 -> +1 = 0

    body = markdown[:line_start].rstrip()
    appendix = markdown[line_start:]
    return body, appendix


def paginate(markdown_body: str, structural_appendix: str = "") -> PaginationResult:
    """
    Splits the body Markdown into virtual pages, never breaking inside a protected
    block (CriticMarkup, tables, footnote sections).

    Args:
        markdown_body: post-projection Markdown WITHOUT the structural appendix.
        structural_appendix: read-only metadata block to append to every page.

    Returns:
        PaginationResult with both finalized PageInfo objects (body+appendix) and
        the pre-injection body_pages with their original-body start offsets.
    """
    # Empty input -> single empty page (or appendix-only page if appendix is non-empty).
    if not markdown_body:
        appendix_clean = structural_appendix.strip() if structural_appendix else ""
        content = appendix_clean
        return PaginationResult(
            pages=[
                PageInfo(
                    page=1,
                    total_pages=1,
                    has_next=False,
                    has_prev=False,
                    tracked_change_count=_count_tracked_changes(content),
                    page_content=content,
                )
            ],
            total_pages=1,
            body_pages=[""],
            body_page_offsets=[0],
        )

    block_records = _tokenize_into_atomic_blocks(markdown_body)
    body_pages, body_page_offsets = _assemble_pages(block_records)

    # Inject appendix on every page.
    if structural_appendix and structural_appendix.strip():
        appendix = structural_appendix.strip()
        final_pages: List[str] = []
        for body_page in body_pages:
            if body_page:
                final_pages.append(f"{body_page}\n\n{appendix}")
            else:
                final_pages.append(appendix)
    else:
        final_pages = list(body_pages)

    total = len(final_pages)
    page_infos = [
        PageInfo(
            page=i,
            total_pages=total,
            has_next=(i < total),
            has_prev=(i > 1),
            tracked_change_count=_count_tracked_changes(content),
            page_content=content,
        )
        for i, content in enumerate(final_pages, start=1)
    ]

    return PaginationResult(
        pages=page_infos,
        total_pages=total,
        body_pages=body_pages,
        body_page_offsets=body_page_offsets,
    )


# ---------------------------------------------------------------------------
# Internal: block tokenization
# ---------------------------------------------------------------------------


def _tokenize_into_atomic_blocks(markdown_body: str) -> List[Tuple[str, int, bool]]:
    """
    Splits markdown_body into atomic blocks. Returns a list of (block_text, start_offset, force_split_after)
    tuples where start_offset is the index in markdown_body where the block begins.

    A block boundary is a "\\n\\n" run where every CriticMarkup wrapper has depth 0.
    Multiple consecutive newlines collapse into a single boundary.

    After raw splitting, post-processing merges:
      a. {==X==}{>>Y<<} pairs joined directly (no whitespace).
         (In practice ingest.py never emits a "\\n\\n" between these — they live in
         the same paragraph — but we guard defensively.)
      b. Footnote/Endnote section headers ("## Footnotes" / "## Endnotes") with all
         subsequent footnote definitions until the next non-footnote block.
    """
    raw_blocks = _split_on_safe_paragraph_breaks(markdown_body)
    merged = _merge_footnote_sections(raw_blocks)

    refined: List[Tuple[str, int, bool]] = []
    for block_text, block_offset in merged:
        if PAGE_BREAK_TOKEN in block_text:
            parts = block_text.split(PAGE_BREAK_TOKEN)
            curr_offset = block_offset
            for idx, part in enumerate(parts):
                part_len = len(part)
                is_last_part = idx == len(parts) - 1
                if part or not is_last_part:
                    refined.append((part, curr_offset, not is_last_part))
                curr_offset += part_len + len(PAGE_BREAK_TOKEN)
        else:
            refined.append((block_text, block_offset, False))

    return refined


def _split_on_safe_paragraph_breaks(text: str) -> List[Tuple[str, int]]:
    """
    Tracks CriticMarkup nesting depth across the string. Splits on "\\n\\n" only
    when all wrapper counters are at zero. Collapses runs of >= 2 newlines into a
    single boundary so we don't emit empty blocks.

    Returns list of (block_text, start_offset) where start_offset is the position
    of block_text in the original `text`.

    PERF: implemented as one compiled-regex scan over the token event stream.
    The historical char-by-char walk called str.startswith at (almost) every
    position — 37 million calls / multiple seconds on a 4.6 MB projection.
    Equivalence with a verbatim copy of that walk is pinned by
    tests/test_pagination_scanner_equivalence.py (crafted, randomized, and
    projection-shaped inputs).
    """
    counters = {close: 0 for close in _CRITIC_TOKENS.values()}
    blocks: List[Tuple[str, int]] = []
    block_start = 0
    n = len(text)

    for m in _SCAN_TOKEN_RE.finditer(text):
        tok = m.group(0)
        first_char = tok[0]
        if first_char == "\n":
            # Paragraph break: a boundary only at wrapper depth zero. The
            # greedy \n{2,} match consumes the whole newline run, mirroring
            # the historical "skip all consecutive newlines" collapse.
            if not any(counters.values()):
                if m.start() > block_start:
                    blocks.append((text[block_start : m.start()], block_start))
                block_start = m.end()
        elif first_char == "{":
            counters[_CRITIC_TOKENS[tok]] += 1
        else:
            # Close token. If unbalanced, it is still consumed (the regex
            # already advanced past it) but never drives a counter negative.
            if counters[tok] > 0:
                counters[tok] -= 1

    # Flush trailing block.
    if block_start < n:
        block_text = text[block_start:n]
        if block_text:
            blocks.append((block_text, block_start))

    return blocks


def _merge_footnote_sections(blocks: List[Tuple[str, int]]) -> List[Tuple[str, int]]:
    """
    Merges a "## Footnotes" or "## Endnotes" header block with all subsequent
    footnote/endnote definition blocks (lines starting with "[^fn-" or "[^en-").

    Footnote definitions are kept attached to their section header so they
    travel together when paginated. Page boundary may still fall *between*
    sections (e.g. body ... | ## Footnotes + defs | ## Endnotes + defs).
    """
    if not blocks:
        return blocks

    merged: List[Tuple[str, int]] = []
    i = 0
    while i < len(blocks):
        block_text, block_offset = blocks[i]
        stripped = block_text.lstrip()
        is_section_header = stripped.startswith("## Footnotes") or stripped.startswith("## Endnotes")

        if not is_section_header:
            merged.append((block_text, block_offset))
            i += 1
            continue

        # Accumulate this header plus all consecutive footnote-definition blocks.
        accumulated_text = block_text
        j = i + 1
        while j < len(blocks):
            next_text, _next_offset = blocks[j]
            next_stripped = next_text.lstrip()
            if next_stripped.startswith("[^fn-") or next_stripped.startswith("[^en-"):
                accumulated_text = f"{accumulated_text}\n\n{next_text}"
                j += 1
            else:
                break

        merged.append((accumulated_text, block_offset))
        i = j

    return merged


# ---------------------------------------------------------------------------
# Internal: page assembly (greedy)
# ---------------------------------------------------------------------------


def _assemble_pages(
    block_records: List[Tuple[str, int, bool]],
) -> Tuple[List[str], List[int]]:
    """
    Greedy bin-packing of blocks into pages of <= PAGE_TARGET_CHARS.

    A block that exceeds PAGE_TARGET_CHARS on its own is emitted as a single
    oversized page (we never split protected content).

    Returns (page_bodies, page_start_offsets) where page_start_offsets[i] is
    the offset in the ORIGINAL markdown_body at which page_bodies[i] begins.
    """
    if not block_records:
        return [""], [0]

    pages: List[str] = []
    page_starts: List[int] = []

    current_blocks: List[str] = []
    current_size = 0
    current_start: int = -1  # offset of first block in the current page

    def flush_current():
        nonlocal current_blocks, current_size, current_start
        if current_blocks:
            non_empty = [b for b in current_blocks if b]
            pages.append("\n\n".join(non_empty) if non_empty else "")
            page_starts.append(current_start)
        current_blocks = []
        current_size = 0
        current_start = -1

    for block_text, block_offset, force_split in block_records:
        block_size = len(block_text)
        added_size = block_size + (2 if current_blocks else 0)  # +2 for "\n\n"

        if current_blocks and current_size + added_size > PAGE_TARGET_CHARS:
            # Adding would overshoot. Flush.
            flush_current()

        if not current_blocks and block_size > PAGE_TARGET_CHARS:
            # Oversized solo block. Emit as its own page.
            pages.append(block_text)
            page_starts.append(block_offset)
            continue

        if not current_blocks:
            current_start = block_offset
        current_blocks.append(block_text)
        current_size += added_size if current_size > 0 else block_size

        if force_split:
            flush_current()

    flush_current()

    if not pages:
        return [""], [0]
    return pages, page_starts


# ---------------------------------------------------------------------------
# Internal: tracked change counting
# ---------------------------------------------------------------------------


def _count_tracked_changes(page_content: str) -> int:
    """Counts distinct Chg:N IDs visible on a page."""
    return len(set(_CHG_ID_PATTERN.findall(page_content)))


# ---------------------------------------------------------------------------
# UI/Formatting: Page banners, footers, and pointers
# ---------------------------------------------------------------------------


def build_page_banner(page: int, total: int, file_path: str, is_cli: bool = False) -> str:
    """
    Returns the top-of-page banner injected into LLM-facing markdown.
    Empty string when there is only one page (no navigation needed).
    """
    if total <= 1:
        return ""
    # "synthetic" is load-bearing: Adeu pages are length-based content chunks
    # sized for LLM consumption, and readers must never mistake them for
    # printed Word pages or explicit page breaks (QA 2026-07-19 ADEU-QA-005).
    if is_cli:
        cmd = f"adeu extract {file_path} --mode outline"
        return (
            f"> **Page {page} of {total}** (synthetic page — a length-based chunk, not a printed "
            f"Word page) — run `{cmd}` for a heading map of the full document.\n\n---\n\n"
        )
    return (
        f"> **Page {page} of {total}** (synthetic page — a length-based chunk, not a printed "
        f"Word page) — call `read_docx` with `mode='outline'` for a heading map of the full document.\n\n"
        f"---\n\n"
    )


def build_page_footer(page: int, total: int, has_next: bool, file_path: str, is_cli: bool = False) -> str:
    """
    Returns the bottom-of-page continuation marker. Empty when this is the
    last page or the only page.
    """
    if total <= 1 or not has_next:
        return ""
    if is_cli:
        cmd = f"adeu extract {file_path} --page {page + 1}"
        return f"\n\n---\n\n> **Continues on page {page + 1} of {total}.** Run `{cmd}` for the next page."
    return f"\n\n---\n\n> **Continues on page {page + 1} of {total}.**"


def build_appendix_pointer(file_path: str, has_appendix: bool, is_cli: bool = False) -> str:
    """
    Returns the one-line footer appended to body pages telling the agent that
    structural metadata (defined terms, cross-references, bookmarks, diagnostics)
    is available via mode='appendix'. Empty string when the document has no
    appendix content.
    """
    if not has_appendix:
        return ""
    if is_cli:
        cmd = f"adeu extract {file_path} --mode appendix --page N"
        return (
            "\n\n---\n\n"
            "> **Appendix available.** This document has structural metadata "
            "(defined terms, cross-references, bookmarks, diagnostics) that may "
            f"be relevant when editing. Run `{cmd}` "
            "to load it before submitting edits."
        )
    return (
        "\n\n---\n\n"
        "> **Appendix available.** This document has structural metadata "
        "(defined terms, cross-references, bookmarks, diagnostics) that may "
        "be relevant when editing. Call `read_docx` with `mode='appendix'` "
        "to load it before submitting edits."
    )

File: node/packages/core/src/pagination.ts
/**
 * Stateless paginator for projected DOCX Markdown.
 */

import { PAGE_BREAK_TOKEN } from './utils/docx.js';

const PAGE_TARGET_CHARS = 19_000;
export const PAGE_RANGE_MAX_PAGES = 8;
const APPENDIX_MARKER = '<!-- READONLY_BOUNDARY_START -->';

export type PageArgKind = "single" | "range" | "all";
const _PAGE_RANGE_RE = /^\s*(\d+)\s*-\s*(\d+)\s*$/;

/** Mirrors python/src/adeu/pagination.py parse_page_arg (:26-89). */
export function parse_page_arg(
  page: number | string | null | undefined,
): [PageArgKind, number | [number, number] | null] {
  const bad = (raw: unknown): never => {
    throw new Error(
      `Invalid page parameter: '${raw}'. Provide a positive integer, page range (e.g. '2-6'), or 'all'.`,
    );
  };
  if (page === null || page === undefined) return ["single", 1];
  if (typeof page === "number") {
    if (!Number.isInteger(page) || page < 1) bad(page);
    return ["single", page];
  }
  if (typeof page === "string") {
    const s = page.trim();
    if (!s) bad(page);
    if (s.toLowerCase() === "all") return ["all", null];
    const m = _PAGE_RANGE_RE.exec(s);
    if (m) {
      const startP = parseInt(m[1], 10);
      const endP = parseInt(m[2], 10);
      if (startP < 1 || endP < 1) bad(page);
      return ["range", [startP, endP]];
    }
    if (!/^\d+$/.test(s)) bad(page);
    const val = parseInt(s, 10);
    if (val < 1) bad(page);
    return ["single", val];
  }
  return bad(page);
}

const _CRITIC_TOKENS: Record<string, string> = {
  '{++': '++}',
  '{--': '--}',
  '{==': '==}',
  '{>>': '<<}',
};

const _CHG_ID_PATTERN = /\bChg:(\d+)\b/g;

export interface PageInfo {
  page: number;
  total_pages: number;
  has_next: boolean;
  has_prev: boolean;
  tracked_change_count: number;
  page_content: string;
}

export interface PaginationResult {
  pages: PageInfo[];
  total_pages: number;
  body_pages: string[];
  body_page_offsets: number[];
}

export function split_structural_appendix(markdown: string): [string, string] {
  if (!markdown) return ['', ''];

  const idx = markdown.indexOf(APPENDIX_MARKER);
  if (idx === -1) return [markdown, ''];

  const line_start = markdown.lastIndexOf('\n', idx) + 1;
  const body = markdown.substring(0, line_start).trimEnd();
  const appendix = markdown.substring(line_start);

  return [body, appendix];
}

export function paginate(markdown_body: string, structural_appendix: string = ''): PaginationResult {
  if (!markdown_body) {
    const appendix_clean = structural_appendix ? structural_appendix.trim() : '';
    const content = appendix_clean;
    return {
      pages: [{
        page: 1,
        total_pages: 1,
        has_next: false,
        has_prev: false,
        tracked_change_count: _count_tracked_changes(content),
        page_content: content,
      }],
      total_pages: 1,
      body_pages: [''],
      body_page_offsets: [0],
    };
  }

  const block_records = _tokenize_into_atomic_blocks(markdown_body);
  const [body_pages, body_page_offsets] = _assemble_pages(block_records);

  let final_pages: string[];
  if (structural_appendix && structural_appendix.trim()) {
    const appendix = structural_appendix.trim();
    final_pages = body_pages.map(bp => bp ? `${bp}\n\n${appendix}` : appendix);
  } else {
    final_pages = [...body_pages];
  }

  const total = final_pages.length;
  const page_infos: PageInfo[] = final_pages.map((content, i) => ({
    page: i + 1,
    total_pages: total,
    has_next: (i + 1 < total),
    has_prev: (i + 1 > 1),
    tracked_change_count: _count_tracked_changes(content),
    page_content: content,
  }));

  return {
    pages: page_infos,
    total_pages: total,
    body_pages,
    body_page_offsets,
  };
}

/**
 * Splits `markdown_body` into atomic blocks as
 * `[block_text, start_offset, force_split_after]`.
 *
 * The third element carries manual page breaks. The projection writes a page
 * break as U+000C (`PAGE_BREAK_TOKEN`), so a block containing one is cut at
 * that point and the preceding part is flagged to end its page — which is how
 * a reader's page numbers line up with Word's. Before this existed the TS port
 * ignored manual breaks entirely and paginated purely by density, so it
 * disagreed with the python engine on every document that used them.
 */
function _tokenize_into_atomic_blocks(
  markdown_body: string,
): [string, number, boolean][] {
  const raw_blocks = _split_on_safe_paragraph_breaks(markdown_body);
  const merged = _merge_footnote_sections(raw_blocks);

  const refined: [string, number, boolean][] = [];
  for (const [block_text, block_offset] of merged) {
    if (block_text.includes(PAGE_BREAK_TOKEN)) {
      const parts = block_text.split(PAGE_BREAK_TOKEN);
      let curr_offset = block_offset;
      for (let idx = 0; idx < parts.length; idx++) {
        const part = parts[idx]!;
        const is_last_part = idx === parts.length - 1;
        if (part || !is_last_part) {
          refined.push([part, curr_offset, !is_last_part]);
        }
        curr_offset += part.length + PAGE_BREAK_TOKEN.length;
      }
    } else {
      refined.push([block_text, block_offset, false]);
    }
  }
  return refined;
}

function _split_on_safe_paragraph_breaks(text: string): [string, number][] {
  const counters: Record<string, number> = { '++}': 0, '--}': 0, '==}': 0, '<<}': 0 };
  const blocks: [string, number][] = [];
  let block_start = 0;
  let i = 0;
  const n = text.length;

  while (i < n) {
    let matched_open = false;
    for (const [open_tok, close_tok] of Object.entries(_CRITIC_TOKENS)) {
      if (text.startsWith(open_tok, i)) {
        counters[close_tok]++;
        i += open_tok.length;
        matched_open = true;
        break;
      }
    }
    if (matched_open) continue;

    let matched_close = false;
    for (const close_tok of Object.values(_CRITIC_TOKENS)) {
      if (text.startsWith(close_tok, i)) {
        if (counters[close_tok] > 0) counters[close_tok]--;
        i += close_tok.length;
        matched_close = true;
        break;
      }
    }
    if (matched_close) continue;

    if (text[i] === '\n' && i + 1 < n && text[i + 1] === '\n') {
      if (Object.values(counters).every(c => c === 0)) {
        const block_text = text.substring(block_start, i);
        if (block_text) blocks.push([block_text, block_start]);

        let j = i;
        while (j < n && text[j] === '\n') j++;
        i = j;
        block_start = i;
        continue;
      }
    }

    i++;
  }

  if (block_start < n) {
    const block_text = text.substring(block_start, n);
    if (block_text) blocks.push([block_text, block_start]);
  }

  return blocks;
}

function _merge_footnote_sections(blocks: [string, number][]): [string, number][] {
  if (!blocks.length) return blocks;

  const merged: [string, number][] = [];
  let i = 0;

  while (i < blocks.length) {
    const [block_text, block_offset] = blocks[i];
    const stripped = block_text.trimStart();
    const is_section_header = stripped.startsWith('## Footnotes') || stripped.startsWith('## Endnotes');

    if (!is_section_header) {
      merged.push([block_text, block_offset]);
      i++;
      continue;
    }

    let accumulated_text = block_text;
    let j = i + 1;
    while (j < blocks.length) {
      const [next_text] = blocks[j];
      const next_stripped = next_text.trimStart();
      if (next_stripped.startsWith('[^fn-') || next_stripped.startsWith('[^en-')) {
        accumulated_text = `${accumulated_text}\n\n${next_text}`;
        j++;
      } else {
        break;
      }
    }

    merged.push([accumulated_text, block_offset]);
    i = j;
  }

  return merged;
}

function _assemble_pages(
  block_records: [string, number, boolean][],
): [string[], number[]] {
  if (!block_records.length) return [[''], [0]];

  const pages: string[] = [];
  const page_starts: number[] = [];

  let current_blocks: string[] = [];
  let current_size = 0;
  let current_start = -1;

  const flush_current = () => {
    if (current_blocks.length > 0) {
      pages.push(current_blocks.join('\n\n'));
      page_starts.push(current_start);
    }
    current_blocks = [];
    current_size = 0;
    current_start = -1;
  };

  for (const [block_text, block_offset, force_split] of block_records) {
    const block_size = block_text.length;
    const added_size = block_size + (current_blocks.length > 0 ? 2 : 0);

    if (current_blocks.length > 0 && current_size + added_size > PAGE_TARGET_CHARS) {
      flush_current();
    }

    if (current_blocks.length === 0 && block_size > PAGE_TARGET_CHARS) {
      pages.push(block_text);
      page_starts.push(block_offset);
      continue;
    }

    if (current_blocks.length === 0) current_start = block_offset;
    current_blocks.push(block_text);
    current_size += current_size > 0 ? added_size : block_size;

    if (force_split) flush_current();
  }

  flush_current();

  if (!pages.length) return [[''], [0]];
  return [pages, page_starts];
}

function _count_tracked_changes(page_content: string): number {
  const matches = [...page_content.matchAll(_CHG_ID_PATTERN)];
  const distinct = new Set(matches.map(m => m[1]));
  return distinct.size;
}
